File: src/lib/rag/context.ts
import type { Candidate, Citation, Conflict, Confidence, IndexedChunk } from "./types";
import { snippet } from "./text";
import { mmrSelect } from "./rerank";

const TOKEN_BUDGET = 2800;

export function assembleContext(cands: Candidate[], k = 8): {
  selected: Candidate[];
  citations: Citation[];
  contradictions: Conflict[];
  confidence: Confidence;
} {
  const diverse = mmrSelect(
    cands.filter((c) => c.rerank >= 0.18),
    Math.min(k, cands.length),
  );
  const selected: Candidate[] = [];
  let tokens = 0;
  for (const c of diverse) {
    if (tokens + c.chunk.tokenCount > TOKEN_BUDGET && selected.length >= 3) break;
    selected.push(c);
    tokens += c.chunk.tokenCount;
  }

  const citations: Citation[] = selected.map((c, i) => ({
    n: i + 1,
    chunkId: c.chunk.id,
    documentId: c.chunk.documentId,
    title: c.chunk.docTitle,
    filename: c.chunk.sourceUri.split("/").pop() ?? c.chunk.docTitle,
    section: c.chunk.section,
    page: c.chunk.page,
    excerpt: snippet(c.chunk.content, 280),
    score: round4(c.rerank),
    validFrom: c.chunk.validFrom,
    validTo: c.chunk.validTo,
    version: c.chunk.version,
  }));

  const contradictions = detectConflicts(selected.map((c) => c.chunk));
  const confidence = scoreConfidence(selected, contradictions);

  return { selected, citations, contradictions, confidence };
}

function detectConflicts(chunks: IndexedChunk[]): Conflict[] {
  const byTopic = new Map<string, IndexedChunk[]>();
  for (const c of chunks) {
    const key = c.collection === "architecture" ? "architecture" : c.documentId;
    const arr = byTopic.get(key) ?? [];
    arr.push(c);
    byTopic.set(key, arr);
  }
  const conflicts: Conflict[] = [];
  const arch = (byTopic.get("architecture") ?? []).filter(
    (c) => /current production architecture/i.test(c.content) || /current platform/i.test(c.section),
  );
  const live = arch.filter((c) => c.status === "indexed");
  const dead = arch.filter((c) => c.status === "deprecated");
  if (live.length && dead.length) {
    const a = dead[0]!;
    const b = live[0]!;
    conflicts.push({
      topic: "Current production architecture",
      a: {
        title: a.docTitle,
        excerpt: snippet(a.content, 180),
        validFrom: a.validFrom,
        version: a.version,
      },
      b: {
        title: b.docTitle,
        excerpt: snippet(b.content, 180),
        validFrom: b.validFrom,
        version: b.version,
      },
      resolution: `${b.docTitle} (${b.validFrom}) supersedes ${a.docTitle} (${a.validFrom}).`,
    });
  }
  return conflicts;
}

function scoreConfidence(selected: Candidate[], conflicts: Conflict[]): Confidence {
  const top = selected[0]?.rerank ?? 0;
  const second = selected[1]?.rerank ?? 0;
  const gap = top - second;
  const coverage = Math.min(1, selected.length / 4);
  const agreement = conflicts.length ? 0.55 : 0.9;
  const density = selected.length
    ? selected.slice(0, 3).reduce((s, c) => s + c.rerank, 0) / Math.min(3, selected.length)
    : 0;
  const raw = 0.4 * top + 0.2 * Math.min(1, gap * 3) + 0.15 * coverage + 0.15 * agreement + 0.1 * density;
  const score = Math.max(0, Math.min(0.99, raw));
  const band =
    score >= 0.9 ? "high" : score >= 0.7 ? "medium" : score >= 0.5 ? "low" : "insufficient";
  return {
    score: round4(score),
    band,
    factors: [
      { name: "top relevance", value: round4(top), note: "Best reranked chunk" },
      { name: "score gap", value: round4(gap), note: "Separation from runner-up" },
      { name: "citation coverage", value: round4(coverage), note: `${selected.length} context chunks` },
      { name: "source agreement", value: round4(agreement), note: conflicts.length ? "conflict detected" : "no conflict" },
      { name: "evidence density", value: round4(density), note: "Mean top-3 rerank" },
    ],
  };
}

function round4(n: number): number {
  return Math.round(n * 10000) / 10000;
}

File: sdk/python/src/aether/context.py
from __future__ import annotations

from .rerank import mmr_select
from .text import round4, snippet
from .types import Candidate, Citation, Confidence, Conflict, IndexedChunk

TOKEN_BUDGET = 2800


def assemble_context(cands: list[Candidate], k: int = 8) -> tuple[list[Candidate], list[Citation], list[Conflict], Confidence]:
    diverse = mmr_select([c for c in cands if c.rerank >= 0.18], min(k, len(cands)))
    selected: list[Candidate] = []
    tokens = 0
    for c in diverse:
        if tokens + c.chunk.token_count > TOKEN_BUDGET and len(selected) >= 3:
            break
        selected.append(c)
        tokens += c.chunk.token_count
    citations = [
        Citation(
            n=i + 1,
            chunk_id=c.chunk.id,
            document_id=c.chunk.document_id,
            title=c.chunk.doc_title,
            filename=c.chunk.source_uri.split("/")[-1] or c.chunk.doc_title,
            section=c.chunk.section,
            page=c.chunk.page,
            excerpt=snippet(c.chunk.content, 280),
            score=round4(c.rerank),
            valid_from=c.chunk.valid_from,
            valid_to=c.chunk.valid_to,
            version=c.chunk.version,
        )
        for i, c in enumerate(selected)
    ]
    contradictions = _detect_conflicts([c.chunk for c in selected])
    confidence = _score_confidence(selected, contradictions)
    return selected, citations, contradictions, confidence


def _detect_conflicts(chunks: list[IndexedChunk]) -> list[Conflict]:
    arch = [
        c
        for c in chunks
        if c.collection == "architecture"
        and (
            re_search(r"current production architecture", c.content)
            or re_search(r"current platform", c.section)
        )
    ]
    live = [c for c in arch if c.status == "indexed"]
    dead = [c for c in arch if c.status == "deprecated"]
    if live and dead:
        a, b = dead[0], live[0]
        return [
            Conflict(
                topic="Current production architecture",
                a={"title": a.doc_title, "excerpt": snippet(a.content, 180), "validFrom": a.valid_from, "version": a.version},
                b={"title": b.doc_title, "excerpt": snippet(b.content, 180), "validFrom": b.valid_from, "version": b.version},
                resolution=f"{b.doc_title} ({b.valid_from}) supersedes {a.doc_title} ({a.valid_from}).",
            )
        ]
    return []


def re_search(pat: str, text: str) -> bool:
    import re

    return bool(re.search(pat, text, re.I))


def _score_confidence(selected: list[Candidate], conflicts: list[Conflict]) -> Confidence:
    top = selected[0].rerank if selected else 0.0
    second = selected[1].rerank if len(selected) > 1 else 0.0
    gap = top - second
    coverage = min(1.0, len(selected) / 4)
    agreement = 0.55 if conflicts else 0.9
    density = (
        sum(c.rerank for c in selected[:3]) / min(3, len(selected)) if selected else 0.0
    )
    raw = 0.4 * top + 0.2 * min(1.0, gap * 3) + 0.15 * coverage + 0.15 * agreement + 0.1 * density
    score = max(0.0, min(0.99, raw))
    band = "high" if score >= 0.9 else "medium" if score >= 0.7 else "low" if score >= 0.5 else "insufficient"
    return Confidence(
        score=round4(score),
        band=band,  # type: ignore[arg-type]
        factors=[
            {"name": "top relevance", "value": round4(top), "note": "Best reranked chunk"},
            {"name": "score gap", "value": round4(gap), "note": "Separation from runner-up"},
            {"name": "citation coverage", "value": round4(coverage), "note": f"{len(selected)} context chunks"},
            {"name": "source agreement", "value": round4(agreement), "note": "conflict detected" if conflicts else "no conflict"},
            {"name": "evidence density", "value": round4(density), "note": "Mean top-3 rerank"},
        ],
    )
